File: Data Structures & Algorithms/add-two-numbers/submission-4.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    ListNode curr;
    int sum;
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(0);
        curr = dummy;

        while (l1 != null && l2 != null) {
            sum += l1.val + l2.val;

            int digit = sum % 10;
            sum -= digit;
            sum /= 10;

            curr.next = new ListNode(digit);

            curr = curr.next;
            l1 = l1.next;
            l2 = l2.next;
        }

        while (l1 != null) {
            sum += l1.val;

            int digit = sum % 10;
            sum -= digit;
            sum /= 10;

            curr.next = new ListNode(digit);

            curr = curr.next;
            l1 = l1.next;
        }

        while (l2 != null) {
            sum += l2.val;

            int digit = sum % 10;
            sum -= digit;
            sum /= 10;

            curr.next = new ListNode(digit);

            curr = curr.next;
            l2 = l2.next;
        }

        if(sum > 0){
            curr.next = new ListNode(sum);
            curr = curr.next;
        }
        
        return dummy.next;
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        min_heap = []

        for curr in lists:
            while curr:
                heapq.heappush(min_heap, curr.val)
                curr = curr.next
        dummy = ListNode(0)
        curr = dummy

        while min_heap:
            curr.next = ListNode(heapq.heappop(min_heap))
            curr = curr.next
        return dummy.next